### PR TITLE
Structure_Engine: Add wall(panel) and floor(panel) and SurfaceElement(panel)

### LINKED ISSUE #2881: Physical_Engine: Floor(Panel) and Wall(Panel) to match IFramingElement(Bar, Usage)

### LINKED ISSUE #2882: Physical.Create.Wall(Construction, ExternalEdges, InternalEdges) ignores InternalEdges.

### PR DESCRIPTION
**Pull request by JosefTaylor**
### Issues addressed by this PR

Closes #2881 
Closes #2882   

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/BHoM_Engine/Structure_Engine/%232883-AddSurfaceElementFromPanel?csf=1&web=1&e=cL8u2i

### Changelog

- Fixed bug in Physical.Create.Wall which ignored internal edges
- Added Wall(Panel), Floor(Panel) and ISurface(Panel, usage) to convert panels to physical surface elements
- Added IConstruction(ISurfaceProperty) to convert surface properties into Physical IConstructions. Convert to maintain total depth and material quantities, but may not accurately reflect geometry or structural properties of the SurfaceProperty. Ribbed properties will contain a void layer to reflect the missing concrete. Thus, it is not a 1:1 translation, until Physical IConstructions can represent the full range of objects represented by ISurfaceProperties.
- Added IVolumePerArea(Construction) matching the IVolumePerArea(SurfaceProperty) which gets the effective thickness of solid materials (important for comparison to MaterialComposition in material takeoff)
- Implemented IVolumePerArea() in SolidVolume() for ISurface and IOpening

### CODE REVIEW

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot fix documentation punctuation ref. 7567525610

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

File: Structure_Engine/Create/Physical/Floor.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/Floor.cs:66]:
```suggestion
                Base.Compute.RecordWarning("The panel has reinforcement, but embedding this information in the physical element is not yet implemented.");
```

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/ISurface.cs:59]: This looks like asking for a crash:
- if `panel.Property == null`
- if `PanelType` property has some unexpected value

I would consider adding more checks

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/ISurface.cs:70]: I would rather record error and return null in those cases, to avoid confusion of someone setting structural usage of all elements to undefined. Alternatively, You could do the check if the panel is horizontal/vertical, then return floor/wall, and error if the panel is sloped.

File: Structure_Engine/Create/Physical/Wall.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/Wall.cs:66]: Same as above, would specify the warning a bit more

File: Structure_Engine/Create/Physical/Wall.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/Wall.cs:47]: This could possibly become a generic method, taken that it is exactly the same as the Floor creation one? But not 200% sure, so feel free to ignore

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [inline Structure_Engine/Query/Construction.cs:64]: Do you know whether the panel layers are ordered by any means? Can imagine insulation landing at the core of the wall 😃

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [inline Structure_Engine/Query/Construction.cs:67]: I would add a warning here - this is definitely a smell good to be aware of

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [inline Structure_Engine/Query/Construction.cs:73]: Are the private methods used anywhere? Can't see a reference?

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [reply Structure_Engine/Create/Physical/ISurface.cs:70]: Structural panels do not have any info on whether they are walls or floors. So the trick would be that you can either:
- specify the surface type (wall/floor/roof etc.)
- let the algorithm decide what is what, with the caveat that anything that is not vertical or horizontal comes out as null, so you need to give it a bit more love before pushing

Not sure how many projects have only orthogonal panels, but would hope such cases could be sped up nicely.

File: Structure_Engine/Create/Physical/Wall.cs
Comment (pawelbaran) [reply Structure_Engine/Create/Physical/Wall.cs:47]: I meant you could try writing a method like `Create.Surface<T>(Panel panel) where T : ISurface`, and depending on type of T you create wall, floor etc.

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [reply Structure_Engine/Query/Construction.cs:67]: Right, something for empirical testing then

File: Physical_Engine/Create/Wall.cs
Comment (JosefTaylor) [reply Physical_Engine/Create/Wall.cs:112]: this brings up an issue that neither of these methods are handling openings properly, I will try to fix.

File: Structure_Engine/Create/Physical/Wall.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/Wall.cs:66]: Agreed

File: Structure_Engine/Query/Construction.cs
Comment (JosefTaylor) [reply Structure_Engine/Query/Construction.cs:73]: Yep, redundant!

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/ISurface.cs:70]: I've made it so that it first looks for a StructuralUsage, and if that's undefined (or null) it looks for the normal. If it's not horizontal or upright, it won't make it (until we have some other physical elements which can be sloped)

**Review by JosefTaylor — COMMENTED** [review]
Made some changes, should resolve your comments @pawelbaran.

**Comment by JosefTaylor** [comment]
@BHoMBot check required

File: Physical_Engine/Create/Wall.cs
Comment (pawelbaran) [inline Physical_Engine/Create/Wall.cs:97]: why do you default `internalEdges` to `null` for walls but not for floors?

File: Physical_Engine/Query/SolidVolume.cs
Comment (pawelbaran) [inline Physical_Engine/Query/SolidVolume.cs:86]: This is a good question... Depends on how we look at the panel: whether we want to get its outer volume or the actual volume of solids included, skipping voids. I do not know the answer to this question tbh

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/ISurface.cs:57]: Wouldn't that work and be safer?
```suggestion
                if (result is StructuralUsage2D)
```

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/ISurface.cs:74]: Please include some tolerance (refers to next line too):
```suggestion
                        if (1 - Math.Abs(panel.Normal().DotProduct(Vector.ZAxis)) <= BH.oM.Geometry.Tolerance.Angle)
```

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [inline Structure_Engine/Create/Physical/ISurface.cs:79]: I would throw an error that some panels could not be identified because not horizontal/vertical, so the user is requested to enforce the type

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [reply Structure_Engine/Query/Construction.cs:64]: Actually now when I look at it, I would not involve `MaterialComposition` in this convert at all. Instead, I would rather work directly on `surfaceProperty.Layers` - it should be simpler plus voids could be maintained easily

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I had another round of review, a few new ideas and thoughts, hopefully driving the PR towards the good 👍 Yet again: the code is obfuscated by some sort of auto formatting, which 1. makes it hard to read 2. I would rather not merge because the code was formatted well before the changes. So it would be great if you could revert those changes (or just start the PR anew) before submitting the final version of the code.

**Comment by JosefTaylor** [comment]
@FraserGreenroyd the bot got confused and tried to run the tests three times!

File: Physical_Engine/Create/Wall.cs
Comment (JosefTaylor) [reply Physical_Engine/Create/Wall.cs:97]: good catch, both should default to null.

File: Physical_Engine/Query/SolidVolume.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/SolidVolume.cs:86]: This aligns with the work we've already done in the Structural space, replacing AverageThickness with VolumePerArea and having a separate method TotalThickness which gets the outer extent of the panel property

This is chiefly useful for LCA work, material takeoff, and visualization, as Revit (for now) is going to ignore the material composition and just find a wall/floor type with the same name, so maintaining both volume (for material takeoff) and thickness (for display purposes) makes sense- I can't think of a better rule without implementing ribbed, waffle, etc. in the Physical space (which should eventually happen)

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/ISurface.cs:57]: Yes, thank you! I was looking for the right way to do this and couldn't find it, and it seemed like the null check and the default case would protect us enough, but this is clearer and probably a bit safer.

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/ISurface.cs:74]: Yeah, I'd considered this and left as-is, but that's the right move.

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/ISurface.cs:79]: agreed

File: Structure_Engine/Query/Construction.cs
Comment (JosefTaylor) [reply Structure_Engine/Query/Construction.cs:64]: That's a good point- using MaterialComposition simplifies a lot for the other surface types, but I had left room for specific methods for types which needed it, and Layered would benefit from maintaining the order of void layers.

**Comment by JosefTaylor** [comment]
@pawelbaran There's no auto-formatting going on, perhaps I'm just too free with the carriage return. I'm not sure what you're seeing as formatting problems, but I don't see why they'd prevent merging.

**Comment by JosefTaylor** [comment]
Ah, no robots involved, just me being messy. is the former (multi line) always preferred, or only when the line is long?

**Comment by FraserGreenroyd** [comment]
> Ah, no robots involved, just me being messy. is the former (multi line) always preferred, or only when the line is long?

Multi-line is preferred when there's more than 1 property being assigned 😄

**Comment by JosefTaylor** [comment]
@FraserGreenroyd @pawelbaran I'm noticing some more formatting that I think Pawel was looking at- it appears BHoMBot decided to remove a bunch of carriage returns and inline a bunch of property sets in [this commit](https://github.com/BHoM/BHoM_Engine/pull/2883/commits/0466db2ca84d589f9b833e0b6c4ce09aaa37b31e)

File: Structure_Engine/Create/Physical/Wall.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/Wall.cs:47]: Tried this, and had some issues with how it came up in Grasshopper- I'd prefer to leave this to a later date, perhaps when more panel types are implemented in Physical oM

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

File: Physical_Engine/Query/MaterialComposition.cs
Comment (michaelhoehn) [reply Physical_Engine/Query/MaterialComposition.cs:204]: After testing, this method still throws an error "1. Cannot evaluate MaterialComposition because the List is empty. "  
With your proposal, you will have 2 messages instead of one to deal with.

**Review by michaelhoehn — CHANGES_REQUESTED** [review]
I'm happy with the overall functionality. I have a couple general questions that I'd like addressed before approval. 

Additionally, please be sure to link your test script for posterity. I made a very basic one and included it in the same folder where yours exists. Please feel free to ignore it.

File: Structure_Engine/Create/Physical/ISurface.cs
Comment (pawelbaran) [reply Structure_Engine/Create/Physical/ISurface.cs]: Such cases always bottom out sooner or later - been there too many times in the geometry engine to let it pass 🙈

File: Structure_Engine/Create/Physical/Wall.cs
Comment (pawelbaran) [reply Structure_Engine/Create/Physical/Wall.cs:48]: Sounds good to me - just in case, my idea was that we would have methods like `Create.Wall` and `Create.Floor` that would have only one line calling a private generic method `Create.ISurface<T>`. So from the UI perspective everything should stay the same. But happy for you to leave as is, actually I am not 100% sure whether my suggested change would be sustainable (thinking of minor differences in how floor or wall gets created)

File: Structure_Engine/Query/Construction.cs
Comment (pawelbaran) [reply Structure_Engine/Query/Construction.cs:84]: Right, I see your point now. I have only worked with constant thickness and layered surfaces, so my perspective was too narrow. However, in this case I would consider dynamic casting with the following implementations:
- dedicated, simple method for constant thickness
- dedicated method for layered
- the method you created for everything else, with a warning saying that ribs etc. are being simplified and voids can be lost

Would that work?

File: Physical_Engine/Query/SolidVolume.cs
Comment (pawelbaran) [reply Physical_Engine/Query/SolidVolume.cs:89]: I would say we are dealing with a philosophical question here, which I personally cannot answer: both concepts of volume are correct, one included nested voids, the other not. So I cannot tell whether this change is right or wrong. Not sure if others can, so I would be fine with merging your proposal under the condition that it will be re-reviewed by @IsakNaslundBh as our main structural philosopher once he is back, before the Beta release. What do you think?

File: Physical_Engine/Query/MaterialComposition.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/MaterialComposition.cs:204]: Hmm, the first error sounds better- this is if you feed an empty list of layers to the Layered SurfaceProperty? If you get the first error, the method should stop and not give you any more.
The second is a warning, as voids will often be included intentionally, but of course sometimes it could be unintentional.

File: Physical_Engine/Query/VolumePerArea.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/VolumePerArea.cs:77]: yeah probably.

File: Physical_Engine/Query/SolidVolume.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/SolidVolume.cs:153]: I hadn't done so as I'm not sure about the other voids, but if that's the desired behavior I'm happy to!

File: Structure_Engine/Create/Physical/Floor.cs
Comment (JosefTaylor) [reply Structure_Engine/Create/Physical/Floor.cs:30]: TBH I'm copying previous work so go eat a bag of something nice.

File: Physical_Engine/Query/MaterialComposition.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/MaterialComposition.cs:204]: I think I need to include a null check before this, so that if none of the layers have a material, it gives up immediately.

File: Physical_Engine/Query/VolumePerArea.cs
Comment (JosefTaylor) [reply Physical_Engine/Query/VolumePerArea.cs:66]: this has the null-check in front of it, which I'd like to use in the other spot as well. Just need to raise an error instead of dumping zero.

File: Structure_Engine/Query/Construction.cs
Comment (JosefTaylor) [reply Structure_Engine/Query/Construction.cs:84]: Yeah, I can add another for constant thickness. The warning comes up when volume < thickness, so I think there's nothing else needed there- I can re-word to specify that the void is a result of ribs/waffles/etc.

File: Structure_Engine/Create/Physical/Floor.cs
Comment (michaelhoehn) [reply Structure_Engine/Create/Physical/Floor.cs:30]: > 🙄 this is fine.

this is fine!

File: Physical_Engine/Query/SolidVolume.cs
Comment (michaelhoehn) [reply Physical_Engine/Query/SolidVolume.cs:153]: I'm happy to leave it as is. This can be picked up at a later date as it's an additional feature.

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Review by michaelhoehn — APPROVED** [review]
I'm happy with the changes. Works as expected. Please make sure to resolve all comments still outstanding prior to check ready merge.

File: Structure_Engine/Query/Construction.cs
Comment (JosefTaylor) [reply Structure_Engine/Query/Construction.cs:116]: Added some additional null-checks around the shop, so the user should see when it happens, and get an idea of what's going on and if it's OK.

**Comment by JosefTaylor** [comment]
@pawelbaran can you accept the changes?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

File: Structure_Engine/Create/Physical/Floor.cs
Comment (pawelbaran) [reply Structure_Engine/Create/Physical/Floor.cs:30]: Is this fine?

Comment (FraserGreenroyd) [reply Structure_Engine/Create/Physical/Floor.cs:30]: image

Comment (michaelhoehn) [reply Structure_Engine/Create/Physical/Floor.cs:30]: > > 🙄 this is fine.
> 
> this is fine!

this is fine?

**Review by pawelbaran — APPROVED** [review]
Happy with the code changes now - glad to have @michaelhoehn to test the functionality as well. The only potential open question is the solid volume, whether it should be area * thickness or area * volume per area. Marking @IsakNaslundBh to hopefully have a look at it and raise concerns if any. Thanks @JosefTaylor well done 👍

**Comment by pawelbaran** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by JosefTaylor** [comment]
@pawelbaran thanks for the final check, do you know what's up with BHoMBot?

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**
